Adding scripts for generating embryonic mouse atlas

pr4deepr opened the pull request:
Generation of a brainglobe compatible Embryonic mouse e15.5 mouse atlas using the data from [Young et al., 2021 paper](https://elifesciences.org/articles/61408) . 
The scripts can be used to generate all the atlases from this paper as long as its in the same format. 
mouse_e15_5.py doesn't have a user interface
mouse_e15_5_UI.py has an interface where the user can enter details for the atlas. This UI is generated using magicgui. 
There is a placeholder for a link to documentation in the UI.

adamltyson commented:
Agree, looks good @pr4deepr. I like the UI version, but if we can reduce the amount of duplicated code (of which we already have a lot of in this repo), that would be great!

pr4deepr commented:
Do you mean that we merge them together?
or generate the UI version so that it calls all the functions from mouse_e15_5.py ?

vigji commented:
Yep! It should actually be factored out of the atlas generation script folder, as ideally that is the folder that we parse for generating automatically the atlases that we upload on G-Node for brainglobe-mediated distribution. As this could be used as the draft for a more general GUI for atlas generation after some additional work to make it more general, maybe it should go to a separate GUI-related folder=
@adamltyson and @FedeClaudi

adamltyson commented:
Yeah I think moving this out to a separate directory is a good idea, this could be a good model for developing a better UI for atlas generation.

pr4deepr commented:
So, would it be a good idea to generate` bg_atlasgen/gui folder`?
and then, add the mouse atlas UI script in there for now?
If I understood this correctly, this UI script would then link to the mouse atlas generation within` bg_atlasgen/atlas_scripts/` ?

I could add a template for a general GUI in there as well, if it helps.

adamltyson commented:
I think that makes sense, @vigji, @FedeClaudi? The GUI code could import any atlas-specific functions from the headless script in the `atlas_scripts` directory.

Any work towards a general GUI would be greatly appreciated, thanks!

vigji commented:
Yes, agreed!

adamltyson commented:
Hey @pr4deepr, just checking in. Do you need any help implementing this? It would be great to get these atlases in BG.

pr4deepr commented:
Apologies, I will get onto this soon. Will keep you updated!

pr4deepr commented:
I've used two libraries magicgui and [magicclass ](https://github.com/hanjinliu/magic-class)for the GUIs. Magicclass has some nice ways of generating multiple buttons using classes and so on. Let me know if the guis work for you or not.

pr4deepr commented:
@adamltyson. I've updated the GUI code. 
BTW, were you asking for the atlases or the GUI code? I may have misunderstood you..

adamltyson commented:
A question for @vigji and @FedeClaudi, based on #16. Now that #17 is merged, how do we want to handle fairly "generic" additional dependencies? @pr4deepr has added the following:
```
PyQt5
SimpleITK
scikit-image
magic-class
```

I can think of two ways we can think about this:
* These dependencies are modern, well maintained packages that are unlikely to have weird dependencies themselves. They are also likely to be used by other atlas scripts, therefore we should keep them in the main `requirements.txt`.
* Even pretty "normal" dependencies could get complicated when we have 50 atlas scripts, so anything new should go as an extra, defined in setup.py.

We could also go for a middle ground, either:
* Put everything into an extra, and incorporate into the main requirements later if the dependency is used my many/most atlases.
* Put the dependencies into the main requirements, and move to an extra if they're causing problems.

Any thoughts? Now I've typed this out, I think I favour: 
> * Put everything into an extra, and incorporate into the main requirements later if the dependency is used my many/most atlases.

It would then be up to the core devs to "tidy" the requirements occasionally.

FedeClaudi commented:
I'm happing with having multiple, potentially repeated sets of dependencies for different atlases. Perhaps if we start seeing some of them repeated across atlases we can add them to the main one (with the exclusion of `PyQt5` due to licensing issues)

adamltyson commented:
Hey @pr4deepr thanks for updating this PR. I've only been able to have a quick look so far, so forgive me if I've missed something. 

I can see two atlas generation GUI scripts, what is the difference between them? One seems to be specific to the embryonic atlas, but the user still needs to supply a URL for the hosted raw data? How does the other GUI (the magicclass one) work? Is it meant to be directly used, and if so, does it only support raw data in a single format?

Also, I tried to run the main atlas script, but I get this, which I guess is easily fixed:
```
  File "/home/adam/Desktop/bg-atlasgen/bg_atlasgen/atlas_scripts/mouse_e15_5.py", line 191, in create_atlas
    RESOLUTION=tuple(map(float, RESOLUTION.strip().split(',')))
AttributeError: 'tuple' object has no attribute 'strip'
```

Thanks for working on this, I think these atlases will be really useful in BG!

adamltyson commented:
> I'm happing with having multiple, potentially repeated sets of dependencies for different atlases. Perhaps if we start seeing some of them repeated across atlases we can add them to the main one (with the exclusion of `PyQt5` due to licensing issues)

Good catch, @pr4deepr would `PySide2` work as a replacement for `PyQt5`? I assume `magicgui` uses `qtpy` anyway?

pr4deepr commented:
> Also, I tried to run the main atlas script, but I get this, which I guess is easily fixed:
> 
> ```
>   File "/home/adam/Desktop/bg-atlasgen/bg_atlasgen/atlas_scripts/mouse_e15_5.py", line 191, in create_atlas
>     RESOLUTION=tuple(map(float, RESOLUTION.strip().split(',')))
> AttributeError: 'tuple' object has no attribute 'strip'
> ```
Ahh, my bad! I've fixed that!


> I can see two atlas generation GUI scripts, what is the difference between them? One seems to be specific to the embryonic atlas, but the user still needs to supply a URL for the hosted raw data? How does the other GUI (the magicclass one) work? Is it meant to be directly used, and if so, does it only support raw data in a single format?

I've updated the code, so there are default entries for each field. It is mainly optimised for the mouse atlases from [Young et al. paper](https://elifesciences.org/articles/61408), but I wasn't sure what needed to modified to make it a general GUI for generating any atlas. 

The GUI scripts differ only on the choice of library for the GUI.  One uses magicgui and other [magicclass](https://hanjinliu.github.io/magic-class/). 
You can generate a gui from functions using magicgui, but I usually have trouble generating different multiple buttons and having menu options using magicgui. 
Magic class is based on magicgui and offers some nice functionality for generating guis from a class, and class methods can be then be used to generate multiple buttons. The magicclass option would be nice if you'd like to expand the gui and include more buttons and menu options.. 


> Good catch, @pr4deepr would `PySide2` work as a replacement for `PyQt5`? I assume `magicgui` uses `qtpy` anyway?

I tried including this in the setup.py file
![image](https://user-images.githubusercontent.com/13831458/143994778-f1ea3765-7da6-4b4e-84d6-55f52c6eed12.png)
and running `python setup.py install `or `pip isntall -e.`
 I get errors that ` "No Qt bindings could be found"` or that some dependencies like `pysgnal `can't be found
I have to `pip install` `magicgui `or `magicclass` to get it working properly. 
Is there another install method?

adamltyson commented:
>Also, I tried to run the main atlas script, but I get this, which I guess is easily fixed:
    `File "/home/adam/Desktop/bg-atlasgen/bg_atlasgen/atlas_scripts/mouse_e15_5.py", line 191, in create_atlas
      RESOLUTION=tuple(map(float, RESOLUTION.strip().split(',')))`
    AttributeError: 'tuple' object has no attribute 'strip'
    Ahh, my bad! I've fixed that!

Thanks!

>I've updated the code, so there are default entries for each field. It is mainly optimised for the mouse atlases from Young et al. paper, but I wasn't sure what needed to modified to make it a general GUI for generating any atlas.

The GUI scripts differ only on the choice of library for the GUI. One uses magicgui and other magicclass.
You can generate a gui from functions using magicgui, but I usually have trouble generating different multiple buttons and having menu options using magicgui.

@FedeClaudi, @vigji what do you think about the GUIs? I like the idea of them, but I'm not sure I want to merge a load of code we're not quite ready to maintain. 

>I tried including this in the setup.py file
image
and running python setup.py install or pip isntall -e.
I get errors that "No Qt bindings could be found" or that some dependencies like pysgnal can't be found
I have to pip install magicgui or magicclass to get it working properly.
Is there another install method?

To merge this, we'd need to move away from the `PyQt5` dependency (due to licensing issues), so you'd need to use `magicgui[pyside2]`. I think the line in `setup.py` should be something like:

```python
extras_require=
    {
    "allenmouse": ["allensdk"]
    "mouse_embryo": ["magicgui[pyside2]", "scikit-image", "magic-class"]
     },
```
Does that work?

FedeClaudi commented:
Yeah I wouldn't want to add extra code to maintain, but if there's an alternative way to still generate the same atlas then we can have both.

In general it would be awesome to have a general GUI for all atlases generations, but that might probe to be too much work for now.

pr4deepr commented:
I've updated the Atlas generation GUI. 
I've tried to make it generalisable to work across Atlases, however the scripts may need modifications.


- The GUI is like this:

![image](https://user-images.githubusercontent.com/13831458/146718193-0f53783c-80cf-4133-9453-8815c1eaa181.png)
- I've added a function to scan the atlas_scripts folder and return names of the atlas generation scripts. It will generate a dropdown menu like this:
![image](https://user-images.githubusercontent.com/13831458/146718284-4c6464b2-421c-46d3-8e8d-6c820a153c9e.png)

- Based on user choice, it will dynamically import the atlas generation script. From each script, the `create_atlas` function will be called. 
- The difference is that the create_atlas function in the atlas generation script will have to accept following parameters:
    - ATLAS_NAME,
    - SPECIES,
    - ATLAS_LINK,
   - ATLAS_FILE_URL,
   -  ORIENTATION,
    - RESOLUTION,
    - CITATION,
    - ROOT_ID,
    - ATLAS_PACKAGER,
    - working_dir

I've  modified the mouse_e_15_5 script to reflect this. 

This script can be used for all of the atlases in the paper: Young et al. 2021, https://doi.org/10.7554/eLife.61408 . So, maybe should rename this to mouse_atlas_young or something similar ??

vigji commented on setup.py:
```diff
@@ -27,4 +27,7 @@
         "Intended Audience :: Science/Research",
     ],
     zip_safe=False,
+    extras_require = {"SimpleITK":["SimpleITK"],
```
First of all, apologies for the very belated review!

If we decide to go with last proposal from @adamltyson this should be something like 

```python
extras_require = {"gui":["SimpleITK", 
                                        "scikit-image"
                                        "magic-class"]}
```
right?

Plus, for the installation to work on a fresh env, I also need to install `magicgui` (maybe that would get automatically installed when installing magic-class?), `pyqtgraph` and  `PyQt5` (@adamltyson installing PySide2 was still giving an import error, it seems something just depend on `PyQt5`).

vigji replied:
After having done all the required installations, I'm stuck because of this error when running the GUI script:

```
File "/path/../anaconda3/envs/test_bgatlasgen/lib/python3.8/site-packages/magicclass/widgets/qtgraph/components.py", line 32, in Region
    changed = Signal(tuple[float, float])
TypeError: 'type' object is not subscriptable
```

@pr4deepr, @adamltyson, any clue where this could be from? I'm on a `conda` fresh env, `Python==3.8`, macOS

pr4deepr replied:
> If you are importing an already existing atlas generation script, what is the advantage of using a GUI? am I missing something here?

Good point! My understanding was to add a GUI or front end for the current atlas generation scripts. 
To have a generic atlas generation script, raw data and all associated files for generating an atlas will need to be quite uniform, i.e., remove the need for atlas-specific generation scripts. Assuming the end user here would be those who generate all the images and annotations for an atlas?

```
extras_require = {"gui":["SimpleITK", 
                                        "scikit-image"
                                        "magic-class"]}
```
Yup, this looks right and magic-class should install magicgui..

I'll have a look at the PyQt issues you're having and get back to you.

vigji replied:
> > If you are importing an already existing atlas generation script, what is the advantage of using a GUI? am I missing something here?
> 
> Good point! My understanding was to add a GUI or front end for the current atlas generation scripts. To have a generic atlas generation script, raw data and all associated files for generating an atlas will need to be quite uniform, i.e., remove the need for atlas-specific generation scripts. Assuming the end user here would be those who generate all the images and annotations for an atlas?

Sorry for not having been clear. Yes, a GUI would also require quite a standardisation of the input files, but the concept would be that maybe that could be achieved by a non-coding user (If it would be a matter of eg saving files in tif format using eg ImageJ, or writing structures hierarchy in a csv file...), and that an atlas generation GUI would pack an atlas provided that the input is standard enough that is can be dealt with by a standard function.

> Yup, this looks right and magic-class should install magicgui..
Ok!

> I'll have a look at the PyQt issues you're having and get back to you.
Great, let me know!

vigji reviewed (commented):
I could not test yet the GUI because of the dependency issue reported, but I have a general conceptual question about the GUI.

@pr4deepr, could you specify a bit better what use case you would envision for your GUI? In my head, the GUI would need no atlas generation script, as it would be targeted to people who can't write their atlas generation script. If you are importing an already existing atlas generation script, what is the advantage of using a GUI? am I missing something here?

adamltyson commented:
Hey @pr4deepr, are you planning on working on this PR further, or shall I close it?

pr4deepr commented:
Hi @adamltyson 
  Thanks for the message. 
Sorry, I'm not working on this at the moment. Don't think I'll get the time to get around to this in the near future. 
Apologies.

Cheers
Pradeep

adamltyson commented:
No worries @pr4deepr. I'll close it for now. If you get some time later on, we'd be very happy to help.

yoda-vid commented:
This looks amazing, @pr4deepr! I ran the GUI for the E15.5 atlas you included, and it generated the atlas and meshes. As I mentioned in your other post, the `tuple out of index` was just from a couple extra parameters that are required (new?). I added settings from one of the other atlases, and it ran smoothly from there.

After perusing the discussion above, am I understanding correctly that the GUI currently requires an underlying custom script but allows the user to apply that script to multiple atlases so long as they are similar and only require a few parameter changes? I imagine that this could work more generally where there are several templates covering most atlas scenarios, and the GUI could allow the user to select a script and populate it with the required parameters non-programmatically (which I think the GUI is close to doing already). Ideally, I suppose the template scripts could ultimately be merged into one pipeline that the GUI would access, bypassing the need for any script selection. From a server standpoint, is there a need to run scripts headlessly? Perhaps the GUI could generate a config file that users would submit to run with the desired template script?

In terms of getting these developing atlases into BrainGlobe, I could imagine a few directions:
1. Copy the E15.5 script for the rest of the atlases, tweaking the required parameters. We could splice out the GUI code for now and handle that in a separate PR.
2. Extend the GUI to generate config files that could be imported into a generic atlas gen script based on your current E15.5 script, with separate config files for each atlas
3. Combo approach, taking the generic gen script with config files for each atlas while deferring the GUI to a separate PR

Let me know how I can help with this! Exciting to see this GUI work.

adamltyson commented:
Hey @pr4deepr, thanks for restarting this effort, it would be great to get these atlases into BrainGlobe!

>From a server standpoint, is there a need to run scripts headlessly?

Yes, we have some automation of atlas packaging & upload, but long-term I see this being very important. If we update the BrainGlobe Atlas API in such a way that requires atlases to be repackaged, we want to be able to run a single script to update them all.

Regarding to the GUI, I think there's some discussion to be had. I had always envisioned a GUI to be very general, and allow non-programmers to package up an atlas. It could guide them through the process, prompting for metadata, and links to specific files as needed. This would then generate a script (or config file) to allow the packaging to be reproduced at any time.  This may be a bit of a pipedream though.

@yoda-vid are all the embryonic atlases structured in a similar way? Could we have a single script that packaged all the atlases up, maybe with a config file pointing to where the files for each atlas are hosted?

yoda-vid commented:
> are all the embryonic atlases structured in a similar way? Could we have a single script that packaged all the atlases up, maybe with a config file pointing to where the files for each atlas are hosted?

@adamltyson yes, they are structured similarly. I am extending @pr4deepr's `mouse_e15_5.py` script to generalize it for the rest of the atlases. @pr4deepr, I'll keep your name as the packager since you did all the work. I'll plan to post it in a new PR while the GUI discussion continues here.

> Regarding to the GUI, I think there's some discussion to be had. I had always envisioned a GUI to be very general, and allow non-programmers to package up an atlas.

I haven't looked through @pr4deepr's GUI closely, but I think it already gets close to this aim. Hopefully we can build on it! It collects most of the relevant metadata, and we should be able to extend it to output a script that you can run non-interactively during atlas updates.

pr4deepr commented:
Thanks @yoda-vid and @adamltyson 

The GUI currently works for mouse atlas from @yoda-vid s paper.
However,  dynamically updating the number of input fields based on the choice of atlas script could be a start as each atlas generation script has slightly different requirements... 
Although, in the longer run a "wizard" style interface would be ideal I think..



> @adamltyson yes, they are structured similarly. I am extending @pr4deepr's `mouse_e15_5.py` script to generalize it for the rest of the atlases. @pr4deepr, I'll keep your name as the packager since you did all the work. I'll plan to post it in a new PR while the GUI discussion continues here.

The advantage of @yoda-vid 's atlases are that they are all uniform, so this one script can be used for all.. Wish all atlases were like this, but of course, there could be domain specific requirements.. :)

adamltyson commented:
Closing this now #25 has been opened.
